### PR TITLE
Add a serviceworker

### PR DESCRIPTION
**Pull request by developit**
Still a few bits to polish, but it works. Same tenchnique we have for workers.

### CODE REVIEW

File: config/auto-sw-plugin/index.js [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    if (this.options.exclude) {
+      const filters = this.createFilter(this.options.exclude);
+      assetMapping = assetMapping.filter(filename => {
+        for (const filter of filters) {
+          if (filter(filename)) return false;
+        }
+        return true;
+      });
+    }
+\    await Promise.all(serviceWorkers.map(
```
Comment (jakearchibald) [inline config/auto-sw-plugin/index.js]: Syntax error

**Review by jakearchibald — CHANGES_REQUESTED** [review]

File: config/auto-sw-plugin/index.js (hunk 1 above)
Comment (developit) [reply config/auto-sw-plugin/index.js]: No idea how this got in there. We don't have an eslint config anymore though, should have caught this.

**Comment by jakearchibald** [comment]
*sigh* I broke the typings rather than fixing them. Will push a fix when I figure it out.

**Review by jakearchibald — APPROVED** [review]
I approve Jason's changes here. Someone else should review mine 😀

**Comment by jakearchibald** [comment]
I'm going to do more UI work around this, but I'll do that against one of the UI branches I already have to avoid conflicts.

**Comment by developit** [comment]
Weird, no idea what `vendor~` would be. It sounds sortof like a chunk name? But we aren't emitting a vendor chunk. FWIW I'm not seeing it.

**Review by surma — APPROVED** [review]
Mostly nits and documentation from my side

File: config/auto-sw-plugin.js [hunk 2]
```diff
@@ -0,0 +1,138 @@
+const util = require('util');
+const minimatch = require('minimatch');
+const SingleEntryPlugin = require('webpack/lib/SingleEntryPlugin');
+const WebWorkerTemplatePlugin = require('webpack/lib/webworker/WebWorkerTemplatePlugin');
+const ParserHelpers = require('webpack/lib/ParserHelpers');
+
+const NAME = 'auto-sw-plugin';
+const JS_TYPES = ['auto', 'esm', 'dynamic'];
+
+module.exports = class AutoSWPlugin {
```
[the remark this answers is not in the document]
Comment (developit) [reply config/auto-sw-plugin.js]: added

File: src/components/compress/index.tsx [hunk 3]
```diff
@@ -153,6 +154,12 @@ async function processSvg(blob: Blob): Promise<HTMLImageElement> {
   return blobToImg(new Blob([newSource], { type: 'image/svg+xml' }));
 }
 
+async function getOldestServiceWorker() {
```
[the remark this answers is not in the document]
Comment (jakearchibald) [reply src/components/compress/index.tsx]: Those are opposites 😄. Would you prefer LeastRecent? I thought "oldest" was an easier way to say that.

File: src/init-app.tsx [hunk 4]
```diff
@@ -9,8 +9,8 @@ let root = document.querySelector('#app') || undefined;
 // "attach" the client-side rendering to it, updating the DOM in-place instead of replacing:
 root = render(<App />, document.body, root);
 
-// In production, this entire condition is removed.
-if (process.env.NODE_ENV === 'development') {
+// In production, install a Service Worker
+if (process.env.NODE_ENV !== 'production') {
```
[the remark this answers is not in the document]
Comment (developit) [reply src/init-app.tsx]: not sure what happened with GH's comments but this line isn't in the current source.

**Comment by jakearchibald** [comment]
Oops, wasn't meant to merge this one without addressing the nits. Will make those changes on master.

**Pull request merged**